### PR TITLE
Remove DynamicallyAccessedMembers on JsonSerializer 

### LINKED ISSUE #52268: Should we remove DynamicallyAccessedMembers attributes on JsonSerializer?
In 5.0, we annotated the JsonSerializer APIs with `DynamicallyAccessedMembers` attributes in #38595.

https://github.com/dotnet/runtime/blob/300c39ddf8a736384418fb01e0338f66c0afff4d/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs#L43-L44

https://github.com/dotnet/runtime/blob/300c39ddf8a736384418fb01e0338f66c0afff4d/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.String.cs#L26-L27

In 6.0, we resolved all the ILLink warnings in System.Text.Json in #51886, which caused us to also add the `RequiresUnreferencedCode` attributes. This is because the trim annotations can't model everything that needs to be preserved to make these APIs trim compatible. For example, see https://github.com/mono/linker/issues/1087.

Having both sets of attributes on these APIs isn't really ideal. The `DynamicallyAccessedMembers` attributes will only preserve members on the top-level type. So having the `DynamicallyAccessedMembers` gives a false sense of hope that this API will work with trimming, when in reality it won't.

The small benefit it does provide is in the case where the user is not serializing a graph, but just a class with primitive properties. However, even in this case, we'd like the user to move to use the source generator instead. So keeping these attributes for this tiny case doesn't seem to outweigh the benefit of the encouragement to move to using the source generator.

We also didn't annotate the DataContractSerializer and XmlSerializer APIs like this. We just marked the unsafe APIs with `RequiresUnreferencedCode`.

However, if we have plans for making the non-source generation Json serializer APIs work with trimming in 6.0, it might not be worth the effort of removing these attributes now, just to add them back in the near future.

cc @vitek-karas @agocke @layomia @steveharter

### PR DESCRIPTION
**Pull request by eerhardt**
Fix #52268

### CODE REVIEW

**Comment by steveharter** [comment]
How are we getting failures like this:
```
CannotRemoveAttribute : Attribute 'System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute' exists on parameter 'returnType' on member 'System.Text.Json.JsonSerializer.Deserialize(System.ReadOnlySpan<System.Byte>, System.Type, System.Text.Json.JsonSerializerOptions)' in the contract but not the implementation.
```

I don't see the attribute in the .ref file.

**Comment by eerhardt** [comment]
> How are we getting failures like this

Ah, it is because we are running API compat on our current "implementation" and comparing it against the previous netcoreapp (i.e. 5.0) "contract". Looks like I'll need to suppress these.

**Review by layomia — APPROVED** [review]
LGTM.

As discussed, we should replace the deleted trimming tests when we have a trim safe option - https://github.com/dotnet/runtime/issues/53393.

Per CI failure it looks like we need to delete the trimming test for `Hashtable`.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/EnumConverterTest.cs [hunk 1]
```diff
@@ -2,15 +2,17 @@
...
 
 namespace SerializerTrimmingTest
 {
     /// <summary>
-    /// Tests that the enum converter factory is linker-safe.
+    /// Tests that the enum converter factory works with trimming.
     /// </summary>
     internal class Program
     {
+        [DynamicDependency(DynamicallyAccessedMemberTypes.PublicProperties, typeof(ClassWithDay))]
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/EnumConverterTest.cs:15]: My thinking was that this test wasn't testing the Serializing of a POCO class, but instead it is explicitly testing that enums work - just like the JsonConverterAttributeTest is just testing that the JsonConverter pointed to by the attribute doesn't get trimmed.

When we convert these tests to use the source generator, this `DynamicDependency` can be removed.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/EnumConverterTest.cs:15]: > When we convert these tests to use the source generator

I opened https://github.com/dotnet/runtime/issues/53437 to track this work.

**Pull request merged**